### PR TITLE
JP-3677: Add maximum_shower_amplitude parameter to jump step

### PR DESCRIPTION
**Pull request by drlaw1558**
This PR addresses JP-3677 by adding a maximum_shower_amplitude parameter to the MIRI cosmic ray showers code, and ensuring that any changes in the rate image due to shower flagging are below this amplitude.

See also corresponding jwst PR https://github.com/spacetelescope/jwst/pull/8890

## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] run regression tests with this branch installed (`"git+https://github.com/<fork>/stcal@<branch>"`)
    - [ ] [`jwst` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) 
    - [ ] [`romancal` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.apichange.rst``: change to public API
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
</details>

### CODE REVIEW

**Comment by drlaw1558** [comment]
Not sure why I can't request a review, but tagging @kmacdonald-stsci

**Review by kmacdonald-stsci — CHANGES_REQUESTED** [review]
There are portions of the code that are invoked unconditionally, but included in conditionals, which results in two parts of the code that need to be maintained, when there could be a single unconditional call to be maintained. 

Also, there are some formatting issues.

File: src/stcal/jump/jump.py [hunk 1]
```diff
@@ -317,6 +320,7 @@ def detect_jumps(
             )
             log.info("Total snowballs = %i", total_snowballs)
             number_extended_events = total_snowballs
+
         if find_showers:
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py]: Done

File: src/stcal/jump/jump.py [hunk 2]
```diff
@@ -1111,6 +1123,39 @@ def find_faint_extended(
                 num_grps_masked=num_grps_masked,
                 max_extended_radius=max_extended_radius
             )
+
+    # Ensure that flagging showers didn't change final fluxes by more than the allowed amount
+    for intg in range(nints):
+        # Approximate pre-shower rates
+        tempdata = indata[intg,:,:,:].copy()
+        # Ignore any groups flagged in the original gdq array
+        tempdata[ingdq[intg,:,:,:] & donotuse_flag != 0] = np.nan
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py]: Done

File: src/stcal/jump/jump.py [hunk 3]
```diff
@@ -1111,6 +1123,39 @@ def find_faint_extended(
                 num_grps_masked=num_grps_masked,
                 max_extended_radius=max_extended_radius
             )
+
+    # Ensure that flagging showers didn't change final fluxes by more than the allowed amount
+    for intg in range(nints):
+        # Approximate pre-shower rates
+        tempdata = indata[intg,:,:,:].copy()
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py]: Done

File: src/stcal/jump/jump.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py]: Done

File: src/stcal/jump/jump.py [hunk 4]
```diff
@@ -1111,6 +1123,39 @@ def find_faint_extended(
...
+        # Compute group differences
+        diff = np.diff(tempdata,axis=0)
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=RuntimeWarning, message="All-NaN")
+            image1 = np.nanmean(diff,axis=0)
+
+        # Approximate post-shower rates
+        tempdata = indata[intg,:,:,:].copy()
+        # Ignore any groups flagged in the shower gdq array
+        tempdata[gdq[intg,:,:,:] & donotuse_flag != 0] = np.nan
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py]: Done

File: src/stcal/jump/jump.py [hunk 5]
```diff
@@ -1111,6 +1123,39 @@ def find_faint_extended(
...
+        diff = np.diff(tempdata,axis=0)
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=RuntimeWarning, message="All-NaN")
+            image2 = np.nanmean(diff,axis=0)
+
+        # Revert the group flags to the pre-shower flags for any pixels whose rates
+        # became NaN or changed by more than the amount reasonable for a real CR shower
+        diff = np.abs(image1 - image2)
+        indx = np.where((np.isfinite(diff) == False)|(diff > max_shower_amplitude))
+        gdq[intg,:,indx[0],indx[1]] = ingdq[intg,:,indx[0],indx[1]]
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py]: Done

File: src/stcal/jump/jump.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py]: Done

File: src/stcal/jump/jump.py [hunk 6]
```diff
@@ -1111,6 +1123,39 @@ def find_faint_extended(
                 num_grps_masked=num_grps_masked,
                 max_extended_radius=max_extended_radius
             )
+
+    # Ensure that flagging showers didn't change final fluxes by more than the allowed amount
+    for intg in range(nints):
```
[the remark this answers is not in the document]
Comment (drlaw1558) [reply src/stcal/jump/jump.py:1089]: Here the worry was that calling .copy() on the entire array could be extremely memory intensive for cases where there were a large number of integrations.  Looping over the integration is less efficient, but it looks like it still only takes 3 seconds even for an extreme TSO case with 257 ints, 30 groups, and 1024x1032 pixels.

**Review by kmacdonald-stsci — APPROVED** [review]
LGTM

**Comment by drlaw1558** [comment]
Updated to catch additional runtime warnings, and ensure the maximum amplitude is passed in the correct units so that it works for both FAST and SLOW mode data.

**Comment by drlaw1558** [comment]
Added one more minor change based on feedback from @mwregan2 to fix a prior bug with propagating flags to smoothed masks.

**Review by tapastro — CHANGES_REQUESTED** [review]
Change fragment needs renamed from 307 --> 306. Otherwise looks good, will approve pending fix.

**Comment by drlaw1558** [comment]
@tapastro Just double checked this for both MIRI MRS and Imaging detectors, and that change is fine.  Without running find_showers the reference pixels do not get flagged as DO_NOT_USE by the jump step, and so this change actually makes their treatment more consistent between whether or not the showers algorithm was run.

**Review by schlafly — APPROVED** [review]

**Review by tapastro — APPROVED** [review]

**Pull request merged**